core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -259,20 +261,21 @@ class Partition(val topic: String,
   }
 
   /**
-   * Check and maybe expand the ISR of the partition.
+   * Check and maybe expand the ISR of the partition. A replica is in ISR of the partition if and only if
+   * replica's LEO >= HW and replica's lag <= replicaLagTimeMaxMs.
    *
    * This function can be triggered when a replica's LEO has incremented
    */
-  def maybeExpandIsr(replicaId: Int) {
+  def maybeExpandIsr(replicaId: Int, logReadResult: LogReadResult) {
     val leaderHWIncremented = inWriteLock(leaderIsrUpdateLock) {
       // check if this replica needs to be added to the ISR
       leaderReplicaIfLocal() match {
         case Some(leaderReplica) =>
           val replica = getReplica(replicaId).get
-          val leaderHW = leaderReplica.highWatermark
           if(!inSyncReplicas.contains(replica) &&
              assignedReplicas.map(_.brokerId).contains(replicaId) &&
-                  replica.logEndOffset.offsetDiff(leaderHW) >= 0) {
+             logReadResult.info.fetchOffsetMetadata.messageOffset >= logReadResult.hw &&
+             logReadResult.fetchTimeMs - replica.lastCaughtUpTimeMs <= replicaManager.config.replicaLagTimeMaxMs) {
```

Comment:
Ah I see. If we do the test provided above, the maybeExpandIsr() will be slightly inaccurate in theory depending on how often maybeShrinkIsr() is executed. Let say replicaLagTimeMaxMs = 10 seconds and maybeShrinkIsr() is executed every 2 seconds. Then a replica in ISR may lag behind leader's LEO by up to 12 seconds, which means that HW may lag behind leader's LEO by up to 12 seconds. Thus a replica may lag behind leader's LEO by up to 12 seconds even if its fetch offset >= hw. If we apply this example to the test above, `logReadResult.fetchTimeMs - replica.lastCaughtUpTimeMs` can be larger than `replicaManager.config.replicaLagTimeMaxMs ` even if `logReadResult.info.fetchOffsetMetadata.messageOffset >= logReadResult.hw`.

Therefore I think this check is nice to have. On the other hand it should be OK to remove that check as well since the 20% inaccuracy is probably not a big deal.